file/file.go

```diff
@@ -111,3 +116,58 @@ func (t *fileTopic) Close() error {
 	}
 	return nil
 }
+
+// --------------------------------------------------------------------
+
+type fileSub struct {
+	root string
+}
+
+// NewSubscriber inits a subscriber within a root directory.
+// It assumes, that file is not being written to anymore.
+// It iterates entire file (all records) without tracking processed messages.
+func NewSubscriber(root string) bps.Subscriber {
+	return &fileSub{
+		root: root,
+	}
+}
+
+func (s *fileSub) Subscribe(ctx context.Context, topic string, handler bps.Handler) error {
+	name := filepath.Join(s.root, topic)
+	f, err := os.Open(name)
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	dec := json.NewDecoder(f)
+	for {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		var rec json.RawMessage
```

Comment:
you can/should decode into `bps.PubMessage`. your `SubMessage` should have a `Data() []byte` but also an `ID() string` message, I think

Reply:
`SubMessage.ID()` can be added later.

Looks like I need to rename `PubMessage` to just `Message` then.

Also... Probably, we don't need a `SubMessage` interface at all?

IMO, just operate `type Message struct {...}` everywhere.

Or you think we can benefit from interface for smth like:

```
type kafka.subMessage *kafka.ConsumerMessage
```

to save some allocs?

(NATS also operates structs https://godoc.org/github.com/nats-io/stan.go#MsgHandler , so it probably makes sense)

Reply:
so we can operate on structs but I still favour a distinction between PubMessage and SubMessage as one will eventually have e.g. an `Ack` method, etc

Reply:
Leaving as is for now, did an alias `type subMessage { PubMessage }` just for `file://` backend.